File: server/services/apple-auth.js
import plist from 'plist';
import fetchCookie from 'fetch-cookie';
import nodeFetch from 'node-fetch';
import os from 'os';
import https from 'https';
import { v4 as uuidv4 } from 'uuid';

const keepAliveAgent = new https.Agent({ keepAlive: true, maxSockets: 20 });
function generateGuid(){
  try{
    const ifs = os.networkInterfaces?.() || {};
    for(const k of Object.keys(ifs)){
      for(const nic of ifs[k] || []){
        const mac = (nic.mac || '').toUpperCase();
        if(mac && mac !== '00:00:00:00:00:00' && mac !== 'FF:FF:FF:FF:FF:FF') return mac.replace(/:/g,'');
      }
    }
  }catch(_){ }
  return uuidv4().replace(/-/g,'').toUpperCase();
}
const RETRY_ERRORS = ['ECONNRESET','EPIPE','ETIMEDOUT'];
const sleep = (ms)=> new Promise(r=>setTimeout(r,ms));
async function requestWithRetry(fetchImpl, url, init={}, retries=3, timeoutMs=25000){
  let last;
  for(let a=0; a<=retries; a++){
    const c = new AbortController();
    const to = setTimeout(()=>c.abort(), timeoutMs);
    try{
      const resp = await fetchImpl(url, { agent: keepAliveAgent, ...init, signal: c.signal });
      clearTimeout(to);
      if(resp.status >= 500 && a < retries){ await sleep(1000*(a+1)); continue; }
      return resp;
    }catch(err){
      clearTimeout(to);
      last = err;
      const msg = String(err?.message || '').toLowerCase();
      const code = err?.code;
      const retriable = msg.includes('socket hang up') || RETRY_ERRORS.includes(code);
      if(retriable && a < retries){ await sleep(1000*(a+1)); continue; }
      throw err;
    }
  }
  throw last;
}
async function postPlistWithRetry(fetchImpl, url, data, headers, retries=3){
  const body = plist.build(data);
  const resp = await requestWithRetry(fetchImpl, url, { method: 'POST', headers, body }, retries);
  const text = await resp.text();
  return plist.parse(text || '<plist></plist>');
}

/**
 * 从 Apple 认证响应中提取认证信息
 * Apple API 可能使用不同的字段名，这个函数会尝试所有可能的字段
 */
function extractAuthInfo(authResponse) {
  if (!authResponse) return null;
  
  // 尝试所有可能的 dsPersonId 字段名
  const dsPersonId = authResponse.dsPersonId || 
                     authResponse.dsPersonID || 
                     authResponse['ds-person-id'] ||
                     authResponse.dsid ||
                     authResponse.personId;
  
  // 尝试所有可能的 passwordToken 字段名
  const passwordToken = authResponse.passwordToken || 
                       authResponse.passwordTokenInfo ||
                       authResponse.tokenInfo?.passwordToken ||
                       authResponse.accountPassword ||
                       authResponse.authToken;
  
  // 其他可能的认证信息
  const displayName = authResponse.displayName;
  const email = authResponse.email || authResponse.appleId;
  
  const authInfo = { dsPersonId, passwordToken, displayName, email };
  
  // 调试：记录提取结果
  const foundFields = Object.entries(authInfo)
    .filter(([key, value]) => value != null)
    .map(([key]) => key);
  
  console.log('[extractAuthInfo] Extracted fields:', foundFields.join(', ') || 'none');
  
  return authInfo;
}

class Store{
  static get guid(){ if(!this._guid) this._guid = generateGuid(); return this._guid; }
  static get Headers(){ return { 'User-Agent':'Configurator/2.15 (Macintosh; OS X 11.0.0; 16G29) AppleWebKit/2603.3.8', 'Content-Type':'application/x-www-form-urlencoded' }; }
  static async authenticate(email, password, mfa){
    const data = { appleId: email, attempt: mfa?2:4, createSession: 'true', guid: this.guid, password: `${password}${mfa??''}`, rmp: 0, why: 'signIn' };
    const url = `https://auth.itunes.apple.com/auth/v1/native/fast?guid=${this.guid}`;
    const parsed = await postPlistWithRetry(this.fetch, url, data, this.Headers, 3);
    
    // 调试：记录认证返回的所有字段
    if (parsed && !parsed.failureType) {
      console.log('[Store.authenticate] Success - Fields:', {
        hasDsPersonId: !!parsed.dsPersonId,
        hasPasswordToken: !!parsed.passwordToken,
        hasDisplayName: !!parsed.displayName,
        hasAccountPassword: !!parsed.accountPassword,
        hasTokenInfo: !!parsed.tokenInfo,
        dsPersonId: parsed.dsPersonId,
        allKeys: Object.keys(parsed)
      });
    }
    
    return { ...parsed, _state: parsed.hasOwnProperty('failureType') ? 'failure' : 'success' };
  }
  static async ensureLicense(appIdentifier, appVerId, Cookie){
    const url = `https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/buyProduct?guid=${this.guid}`;
    const data = { guid: this.guid, salableAdamId: appIdentifier, ...(appVerId && { externalVersionId: appVerId, appExtVrsId: appVerId }), pricingParameters: 'STDQ' };
    
    // 使用辅助函数提取认证信息
    const authInfo = extractAuthInfo(Cookie);
    
    // 调试：记录传递的认证信息
    console.log('[Store.ensureLicense] Cookie info:', {
      hasDsPersonId: !!authInfo?.dsPersonId,
      hasPasswordToken: !!authInfo?.passwordToken,
      dsPersonId: authInfo?.dsPersonId,
      originalCookieKeys: Cookie ? Object.keys(Cookie) : 'null'
    });
    
    const headers = { 
      ...this.Headers, 
      'X-Dsid': authInfo?.dsPersonId, 
      'iCloud-DSID': authInfo?.dsPersonId, 
      ...(authInfo?.passwordToken ? { 'X-Token': authInfo.passwordToken } : {}) 
    };
    return await postPlistWithRetry(this.fetch, url, data, headers, 3);
  }
  static async downloadProduct(appIdentifier, appVerId, Cookie){
    const url = `https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/volumeStoreDownloadProduct?guid=${this.guid}`;
    const data = { creditDisplay: '', guid: this.guid, salableAdamId: appIdentifier, ...(appVerId && { externalVersionId: appVerId }) };
    
    // 使用辅助函数提取认证信息
    const authInfo = extractAuthInfo(Cookie);
    
    // 调试：记录传递的认证信息
    console.log('[Store.downloadProduct] Cookie info:', {
      hasDsPersonId: !!authInfo?.dsPersonId,
      hasPasswordToken: !!authInfo?.passwordToken,
      dsPersonId: authInfo?.dsPersonId,
      originalCookieKeys: Cookie ? Object.keys(Cookie) : 'null'
    });
    
    const headers = { 
      ...this.Headers, 
      'X-Dsid': authInfo?.dsPersonId, 
      'iCloud-DSID': authInfo?.dsPersonId, 
      ...(authInfo?.passwordToken ? { 'X-Token': authInfo.passwordToken } : {}) 
    };
    let parsed = await postPlistWithRetry(this.fetch, url, data, headers, 3);
    
    // 调试：记录下载结果
    console.log('[Store.downloadProduct] Result:', {
      hasFailureType: !!parsed?.failureType,
      failureType: parsed?.failureType,
      customerMessage: parsed?.customerMessage,
      state: parsed?._state,
      hasSongList: !!parsed?.songList
    });
    
    // 更全面的 license 错误检测
    const errorMsg = String(parsed?.failureType || parsed?.customerMessage || parsed?.message || '').toLowerCase();
    const isLicenseError = /license|not found|未购买|未找到|未授权|not purchased/i.test(errorMsg);
    
    if (parsed?.failureType && isLicenseError){
      await this.ensureLicense(appIdentifier, appVerId, Cookie);
      parsed = await postPlistWithRetry(this.fetch, url, data, headers, 3);
    }
    return { ...parsed, _state: parsed.hasOwnProperty('failureType') ? 'failure' : 'success' };
  }
}
Store.cookieJar = new fetchCookie.toughCookie.CookieJar();
Store.fetch = fetchCookie(nodeFetch, Store.cookieJar);
export { Store };

class AccountStore{
  constructor(){ this.cookieJar = new fetchCookie.toughCookie.CookieJar(); this.fetch = fetchCookie(nodeFetch, this.cookieJar); this._guid = generateGuid(); this.Headers = { 'User-Agent':'Configurator/2.15 (Macintosh; OS X 11.0.0; 16G29) AppleWebKit/2603.3.8', 'Content-Type':'application/x-www-form-urlencoded' }; }
  get guid(){ return this._guid; }
  async authenticate(email, password, mfa){
    const data = { appleId: email, attempt: mfa?2:4, createSession:'true', guid:this.guid, password:`${password}${mfa??''}`, rmp:0, why:'signIn' };
    const url = `https://auth.itunes.apple.com/auth/v1/native/fast?guid=${this.guid}`;
    const parsed = await postPlistWithRetry(this.fetch, url, data, this.Headers, 3);
    
    // 调试：记录认证返回的所有字段
    if (parsed && !parsed.failureType) {
      console.log('[AccountStore.authenticate] Success - Fields:', {
        hasDsPersonId: !!parsed.dsPersonId,
        hasPasswordToken: !!parsed.passwordToken,
        hasDisplayName: !!parsed.displayName,
        hasAccountPassword: !!parsed.accountPassword,
        hasTokenInfo: !!parsed.tokenInfo,
        dsPersonId: parsed.dsPersonId,
        allKeys: Object.keys(parsed)
      });
    }
    
    return { ...parsed, _state: parsed.hasOwnProperty('failureType') ? 'failure' : 'success' };
  }
  async ensureLicense(appIdentifier, appVerId, Cookie){
    const url = `https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/buyProduct?guid=${this.guid}`;
    const data = { guid:this.guid, salableAdamId:appIdentifier, ...(appVerId && { externalVersionId:appVerId, appExtVrsId:appVerId }), pricingParameters:'STDQ' };
    
    // 使用辅助函数提取认证信息
    const authInfo = extractAuthInfo(Cookie);
    
    // 调试：记录传递的认证信息
    console.log('[AccountStore.ensureLicense] Cookie info:', {
      hasDsPersonId: !!authInfo?.dsPersonId,
      hasPasswordToken: !!authInfo?.passwordToken,
      dsPersonId: authInfo?.dsPersonId,
      originalCookieKeys: Cookie ? Object.keys(Cookie) : 'null'
    });
    
    const headers = { 
      ...this.Headers, 
      'X-Dsid': authInfo?.dsPersonId, 
      'iCloud-DSID': authInfo?.dsPersonId, 
      ...(authInfo?.passwordToken ? { 'X-Token': authInfo.passwordToken } : {}) 
    };
    return await postPlistWithRetry(this.fetch, url, data, headers, 3);
  }
  async downloadProduct(appIdentifier, appVerId, Cookie){
    const url = `https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/volumeStoreDownloadProduct?guid=${this.guid}`;
    const data = { creditDisplay:'', guid:this.guid, salableAdamId:appIdentifier, ...(appVerId && { externalVersionId:appVerId }) };
    
    // 使用辅助函数提取认证信息
    const authInfo = extractAuthInfo(Cookie);
    
    // 调试：记录传递的认证信息
    console.log('[AccountStore.downloadProduct] Cookie info:', {
      hasDsPersonId: !!authInfo?.dsPersonId,
      hasPasswordToken: !!authInfo?.passwordToken,
      dsPersonId: authInfo?.dsPersonId,
      originalCookieKeys: Cookie ? Object.keys(Cookie) : 'null'
    });
    
    const headers = { 
      ...this.Headers, 
      'X-Dsid': authInfo?.dsPersonId, 
      'iCloud-DSID': authInfo?.dsPersonId, 
      ...(authInfo?.passwordToken ? { 'X-Token': authInfo.passwordToken } : {}) 
    };
    let parsed = await postPlistWithRetry(this.fetch, url, data, headers, 3);
    
    // 调试：记录下载结果
    console.log('[AccountStore.downloadProduct] Result:', {
      hasFailureType: !!parsed?.failureType,
      failureType: parsed?.failureType,
      customerMessage: parsed?.customerMessage,
      state: parsed?._state,
      hasSongList: !!parsed?.songList
    });
    
    // 更全面的 license 错误检测
    const errorMsg = String(parsed?.failureType || parsed?.customerMessage || parsed?.message || '').toLowerCase();
    const isLicenseError = /license|not found|未购买|未找到|未授权|not purchased/i.test(errorMsg);
    
    if (parsed?.failureType && isLicenseError){
      await this.ensureLicense(appIdentifier, appVerId, Cookie);
      parsed = await postPlistWithRetry(this.fetch, url, data, headers, 3);
    }
    return { ...parsed, _state: parsed.hasOwnProperty('failureType') ? 'failure' : 'success' };
  }
}
export { AccountStore };

File: server/src/apple_auth.rs
use reqwest::{header, Client};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::time::Duration;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthInfo {
    pub ds_person_id: Option<String>,
    pub password_token: Option<String>,
    pub display_name: Option<String>,
    pub email: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Store {
    pub client: Client,
    pub guid: String,
}

impl Store {
    pub fn new() -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .unwrap();

        // 生成 GUID（使用 MAC 地址或随机 UUID）
        let guid = Self::generate_guid();

        Store { client, guid }
    }

    fn generate_guid() -> String {
        // 简单的 GUID 生成
        uuid::Uuid::new_v4()
            .to_string()
            .to_uppercase()
            .replace("-", "")
    }

    pub fn get_headers() -> header::HeaderMap {
        let mut headers = header::HeaderMap::new();
        headers.insert(
            "User-Agent",
            "Configurator/2.15 (Macintosh; OS X 11.0.0; 16G29) AppleWebKit/2603.3.8"
                .parse()
                .unwrap(),
        );
        headers.insert(
            "Content-Type",
            "application/x-www-form-urlencoded".parse().unwrap(),
        );
        headers
    }

    pub async fn authenticate(
        &self,
        email: &str,
        password: &str,
        mfa: Option<&str>,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!(
            "https://auth.itunes.apple.com/auth/v1/native/fast?guid={}",
            self.guid
        );

        let mut auth_data = HashMap::new();
        auth_data.insert("appleId", Value::String(email.to_string()));
        auth_data.insert(
            "attempt",
            Value::Number(serde_json::Number::from(if mfa.is_some() { 2 } else { 4 })),
        );
        auth_data.insert("createSession", Value::String("true".to_string()));
        auth_data.insert("guid", Value::String(self.guid.clone()));
        auth_data.insert(
            "password",
            Value::String(format!("{}{}", password, mfa.unwrap_or(""))),
        );
        auth_data.insert("rmp", Value::Number(serde_json::Number::from(0)));
        auth_data.insert("why", Value::String("signIn".to_string()));

        // 将数据转换为 plist 格式（这里简化处理）
        let response = self
            .client
            .post(&url)
            .headers(Self::get_headers())
            .form(&auth_data)
            .send()
            .await?;

        let result: HashMap<String, Value> = response.json().await?;

        let mut final_result = result.clone();
        if result.contains_key("failureType") {
            final_result.insert("_state".to_string(), Value::String("failure".to_string()));
        } else {
            final_result.insert("_state".to_string(), Value::String("success".to_string()));
        }

        Ok(final_result)
    }

    pub async fn ensure_license(
        &self,
        app_identifier: &str,
        app_ver_id: Option<&str>,
        auth_info: &AuthInfo,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!(
            "https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/buyProduct?guid={}",
            self.guid
        );

        let mut purchase_data = HashMap::new();
        purchase_data.insert("guid", Value::String(self.guid.clone()));
        purchase_data.insert("salableAdamId", Value::String(app_identifier.to_string()));
        if let Some(ver_id) = app_ver_id {
            purchase_data.insert("externalVersionId", Value::String(ver_id.to_string()));
            purchase_data.insert("appExtVrsId", Value::String(ver_id.to_string()));
        }
        purchase_data.insert("pricingParameters", Value::String("STDQ".to_string()));

        let mut headers = Self::get_headers();
        if let Some(ds_id) = &auth_info.ds_person_id {
            headers.insert("X-Dsid", ds_id.parse().unwrap());
            headers.insert("iCloud-DSID", ds_id.parse().unwrap());
        }
        if let Some(token) = &auth_info.password_token {
            headers.insert("X-Token", token.parse().unwrap());
        }

        let response = self
            .client
            .post(&url)
            .headers(headers)
            .form(&purchase_data)
            .send()
            .await?;

        let mut result: HashMap<String, Value> = response.json().await?;
        result.insert("_state".to_string(), Value::String("success".to_string()));
        Ok(result)
    }

    pub async fn download_product(
        &self,
        app_identifier: &str,
        app_ver_id: Option<&str>,
        auth_info: &AuthInfo,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!(
            "https://p25-buy.itunes.apple.com/WebObjects/MZFinance.woa/wa/volumeStoreDownloadProduct?guid={}",
            self.guid
        );

        let mut download_data = HashMap::new();
        download_data.insert("creditDisplay", Value::String("".to_string()));
        download_data.insert("guid", Value::String(self.guid.clone()));
        download_data.insert("salableAdamId", Value::String(app_identifier.to_string()));
        if let Some(ver_id) = app_ver_id {
            download_data.insert("externalVersionId", Value::String(ver_id.to_string()));
        }

        let mut headers = Self::get_headers();
        if let Some(ds_id) = &auth_info.ds_person_id {
            headers.insert("X-Dsid", ds_id.parse().unwrap());
            headers.insert("iCloud-DSID", ds_id.parse().unwrap());
        }
        if let Some(token) = &auth_info.password_token {
            headers.insert("X-Token", token.parse().unwrap());
        }

        let response = self
            .client
            .post(&url)
            .headers(headers)
            .form(&download_data)
            .send()
            .await?;

        let mut result: HashMap<String, Value> = response.json().await?;
        result.insert("_state".to_string(), Value::String("success".to_string()));
        Ok(result)
    }
}

impl Default for Store {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone)]
pub struct AccountStore {
    pub store: Store,
    pub account_email: String,
    pub auth_info: Option<AuthInfo>,
}

impl AccountStore {
    pub fn new(email: &str) -> Self {
        AccountStore {
            store: Store::new(),
            account_email: email.to_string(),
            auth_info: None,
        }
    }

    pub async fn authenticate(
        &mut self,
        password: &str,
        mfa: Option<&str>,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let result = self
            .store
            .authenticate(&self.account_email, password, mfa)
            .await?;

        // 提取认证信息
        if result.get("_state").and_then(|v| v.as_str()) == Some("success") {
            let auth_info = AuthInfo {
                ds_person_id: result
                    .get("dsPersonId")
                    .and_then(|v| v.as_str())
                    .map(String::from),
                password_token: result
                    .get("passwordToken")
                    .and_then(|v| v.as_str())
                    .map(String::from),
                display_name: result
                    .get("displayName")
                    .and_then(|v| v.as_str())
                    .map(String::from),
                email: Some(self.account_email.clone()),
            };
            self.auth_info = Some(auth_info);
        }

        Ok(result)
    }

    pub async fn download_product(
        &self,
        app_identifier: &str,
        app_ver_id: Option<&str>,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let auth_info = self.auth_info.as_ref().ok_or("Not authenticated")?;
        self.store
            .download_product(app_identifier, app_ver_id, auth_info)
            .await
    }

    pub async fn ensure_license(
        &self,
        app_identifier: &str,
        app_ver_id: Option<&str>,
    ) -> Result<HashMap<String, Value>, Box<dyn std::error::Error + Send + Sync>> {
        let auth_info = self.auth_info.as_ref().ok_or("Not authenticated")?;
        self.store
            .ensure_license(app_identifier, app_ver_id, auth_info)
            .await
    }
}
